lib/sda/service_test.go

```diff
@@ -0,0 +1,485 @@
+package sda_test
+
+import (
+	"github.com/dedis/cothority/lib/dbg"
+	"github.com/dedis/cothority/lib/network"
+	"github.com/dedis/cothority/lib/sda"
+	"github.com/satori/go.uuid"
+	"testing"
+	"time"
+)
+
+type DummyProtocol struct {
+	*sda.TreeNodeInstance
+	link   chan bool
+	config DummyConfig
+}
+
+type DummyConfig struct {
+	A    int
+	Send bool
+}
+
+type DummyMsg struct {
+	A int
+}
+
+var dummyMsgType network.MessageTypeID
+
+func init() {
+	dummyMsgType = network.RegisterMessageType(DummyMsg{})
+}
+
+func NewDummyProtocol(tni *sda.TreeNodeInstance, conf DummyConfig, link chan bool) *DummyProtocol {
+	return &DummyProtocol{tni, link, conf}
+}
+
+func (dm *DummyProtocol) Start() error {
+	dm.link <- true
+	if dm.config.Send {
+		if err := dm.SendTo(dm.TreeNode(), &DummyMsg{}); err != nil {
+			dbg.Error(err)
+		}
+		// also send to the children if any
+		if !dm.IsLeaf() {
+			if err := dm.SendTo(dm.Children()[0], &DummyMsg{}); err != nil {
+				dbg.Error(err)
+			}
+		}
+	}
+	return nil
+}
+
+func (dm *DummyProtocol) DispatchMsg(msg *sda.Data) {
+	dm.link <- true
+}
+
+// legcy reasons
+func (dm *DummyProtocol) Dispatch() error {
+	return nil
+}
+
+type DummyService struct {
+	c        sda.Context
+	path     string
+	link     chan bool
+	fakeTree *sda.Tree
+	firstTni *sda.TreeNodeInstance
+	Config   DummyConfig
+}
+
+func (ds *DummyService) ProcessRequest(e *network.Entity, r *sda.Request) {
+	if r.Type != dummyMsgType {
+		ds.link <- false
+		return
+	}
+	if ds.firstTni == nil {
+		ds.firstTni = ds.c.NewTreeNodeInstance(ds.fakeTree, ds.fakeTree.Root)
+	}
+
+	dp := NewDummyProtocol(ds.firstTni, ds.Config, ds.link)
+
+	if err := ds.c.RegisterProtocolInstance(dp); err != nil {
+		ds.link <- false
+		return
+	}
+	dp.Start()
+}
+
+func (ds *DummyService) NewProtocol(tn *sda.TreeNodeInstance, conf *sda.GenericConfig) (sda.ProtocolInstance, error) {
+	dp := NewDummyProtocol(tn, DummyConfig{}, ds.link)
+	return dp, nil
+}
+
+func TestServiceNew(t *testing.T) {
+	defer dbg.AfterTest(t)
+	dbg.TestOutput(testing.Verbose(), 4)
+	ds := &DummyService{
+		link: make(chan bool),
+	}
+	sda.RegisterNewService("DummyService", func(c sda.Context, path string) sda.Service {
+		ds.c = c
+		ds.path = path
+		ds.link <- true
+		return ds
+	})
+	go func() {
+		h := sda.NewLocalHost(2000)
+		h.Close()
+	}()
+
+	waitOrFatal(ds.link, t)
+}
+
+func TestServiceProcessRequest(t *testing.T) {
+	defer dbg.AfterTest(t)
+	dbg.TestOutput(testing.Verbose(), 4)
+	ds := &DummyService{
+		link: make(chan bool),
+	}
+	sda.RegisterNewService("DummyService", func(c sda.Context, path string) sda.Service {
+		ds.c = c
+		ds.path = path
+		return ds
+	})
+	host := sda.NewLocalHost(2000)
+	host.Listen()
+	host.StartProcessMessages()
+	dbg.Lvl1("Host created and listening")
+	defer host.Close()
+	// Send a request to the service
+	re := &sda.Request{
+		Service: sda.ServiceFactory.ServiceID("DummyService"),
+		Type:    network.MessageTypeID(uuid.Nil),
+	}
+	// fake a client
```

Comment:
I leave it for future discussion, I mainly focus on having this PR mergeable and usable, and future improvements like this can be done in later PR (and we'll surely do what you propose since we can basically copy and paste, while the from_scratch is advancing..)